File: jotupix_demo_windows/jotupix/core/JPacket.cpp
//#include "pch.h"
#include <string.h>
#include "JPacket.h"
#include "JByteWriter.h"

#define TAG  "JPacket"

int JPacket::Send(JPacket::Data* psPktData, JProtocol* pProtocol)
{
    uint8_t u8CheckSum = 0;
    uint16_t i = 0;
    JByteWriter sendData;

    if (psPktData == NULL || pProtocol == NULL)
    {
        JLogE(TAG, "Send fail\r\n");
        return -1;
    }

    // The first byte is the message type
    sendData.put_u8(psPktData->m_u8MsgType);

    // packet type, default 0
    uint8_t u8Type = 0;
    if (psPktData->m_ePacketTransType == TransType::FLAG)
    {
        u8Type |= 0x01;
    }

    if (psPktData->m_u8TransCompleteFlag > 0)
    {
        u8Type |= (0x01 << 1);
    }

    sendData.put_u8(u8Type);

    // all data len
    sendData.put_u32(psPktData->m_u32AllDataLen);

    // packet id
    sendData.put_u16(psPktData->m_u16PacketId);

    // Current packet data length, 2 bytes, subsequent padding
    sendData.put_u16(psPktData->m_u16PaylaodLen);

    sendData.put_bytes(psPktData->m_pu8Payload, psPktData->m_u16PaylaodLen);

    // Get the check byte, excluding the message type
    for (i = 1; i < sendData.size(); i++)
    {
        u8CheckSum ^= sendData.buffer[i];
    }

    sendData.put_u8(u8CheckSum);

    return pProtocol->Send(sendData.buffer.data(), sendData.size());
}

File: jotupix_demo_ios/JotupixDemo/Jotupix/Core/JPacket.m
// JPacket.m
#import "JPacket.h"
#import "JByteWriter.h"
#import "JLog.h"

#define TAG "JPacket"

@implementation JPacket

/// Build and send a packet through JProtocol.
- (int)send:(JPacketData *)pkt protocol:(JProtocol *)protocol
{
    // Validate parameters
    if (!pkt || !protocol) {
        JLogE(TAG, "Send fail\r\n");
        return -1;
    }

    uint8_t checksum = 0;   // Checksum accumulator
    uint16_t i = 0;

    JByteWriter *writer = [[JByteWriter alloc] init];

    // 1. Write message type
    [writer put_u8:pkt->msgType];

    // 2. Encode transmission flags (packet type bits)
    uint8_t type = 0;
    
    // FLAG transmission method
    if (pkt->transType == JPacketTransType_FLAG)
        type |= 0x01;

    // Completed-flag bit
    if (pkt->transCompleteFlag > 0)
        type |= (0x01 << 1);

    [writer put_u8:type];

    // 3. Total data length (used in LEN mode)
    [writer put_u32:pkt->allDataLen];

    // 4. Packet sequence ID
    [writer put_u16:pkt->packetId];

    // 5. Current packet payload length
    [writer put_u16:pkt->payloadLen];

    // 6. Write payload bytes
    if (pkt->payload && pkt->payloadLen > 0) {
        [writer put_bytes:pkt->payload length:pkt->payloadLen];
    }

    // 7. Compute checksum (XOR of all bytes except the first one)
    for (i = 1; i < writer.size; i++) {
        checksum ^= writer.buffer[i];
    }

    [writer put_u8:checksum];

    // 8. Send constructed packet via protocol 
    return [protocol send:writer.buffer length:writer.size];
}

@end
